### ISSUE TITLE
openvkl tests cant find libraries on windows. 

### ISSUE THREAD

**Issue by semeraro** [openvkl/openvkl#21]
I am trying to build and use openvkl on a windows system (forgive me father... ). I got the superbuild to complete but when I try to run one of the examples in the bin directory I get 

>C:\Users\Dave Semeraro>vklMinimal_03.exe
>[openvkl] Load of openvkl_module_cpu_device_4 failed due to: 'could not open module lib >openvkl_module_cpu_device_4: The specified module could not be found.
>'
>[openvkl] Load of openvkl_module_cpu_device_8 failed due to: 'could not open module lib >openvkl_module_cpu_device_8: The specified module could not be found.
>'
>[openvkl] Load of openvkl_module_cpu_device_16 failed due to: 'could not open module lib >openvkl_module_cpu_device_16: The specified module could not be found.
>'
>[openvkl] INITIALIZATION ERROR: Could not find device of type: cpu_8.  Make sure you have the >correct VKL libraries linked.

I have added the full path to the device dll files to my windows path environment but still no luck. Setting the Path works for the tbb and rkcommon dll files that are also needed by the openvkl examples but not for the cpu device itself. I am running on windows 11 with v msvc 2022 release. Any windows gurus out there? 

This isnt really a bug but I didnt know where else to ask the question.

**Comment by kraszkow** [comment]
Hi @semeraro, which version of OpenVKL are you using (which branch/tag/commit sha)?

**Comment by semeraro** [comment]
I cloned the master branch of version 1.3.2 from github. Tag 15 I suppose. I just pulled it down two days ago. I suspect it has something to do with rkcommon which I believe handles loading of the dlls. I did the superbuild straight out of the box. Release build type and left the cmake settings as they were.

Dave
________________________________

**Comment by semeraro** [comment]
The device files and openvkl.dll are all placed in install/openvkl/bin. They are all in the same directory on my build. The openvkl_module_cpu_device dll files still dont load.

**Comment by kraszkow** [comment]
Thanks Dave. So it's something with how you're building it. Could you attach here binaries (all dlls) which you've built? 
I will try to narrow down this issue.

**Comment by semeraro** [comment]
I can do that. There are lots of them including all the dependencies. Are you sure you want all of them or just the openvkl stuff?

**Comment by kraszkow** [comment]
openvkl*.dll + rkcommon.dll should be sufficient.

**Comment by semeraro** [comment]
how do I attach dll files. File type not supported errors.

**Comment by kraszkow** [comment]
zip them? and then follow: https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/attaching-files

**Comment by semeraro** [comment]
[openvkl_module_cpu_device_16.zip](https://github.com/openvkl/openvkl/files/11435722/openvkl_module_cpu_device_16.zip)
the openvkl bits.

**Comment by semeraro** [comment]
thanks. I have cloned a fresh copy of rkcommon. Configure and built with no cmake changes. Failed all of the ctests.

**Comment by kraszkow** [comment]
Thanks, but I can't reproduce this problem. My steps:

1) Download & unzip https://github.com/openvkl/openvkl/releases/download/v1.3.2/openvkl-1.3.2.x86_64.windows.zip
2) Download & unzip & overwrite rkcommon.dll & openvkl*.dll
3) Run examples, tutorial etc. - all works fine on my side:

![image](https://github.com/openvkl/openvkl/assets/31285540/55ddce6d-e76e-41e8-8a1a-550f43f2fd79)


If above steps can give you same initial error then most likely only difference is that I'm using Win10 (vs win11 you're using).

**Comment by semeraro** [comment]
Interesting. Ill give that a try and see what happens on my side.

**Comment by kraszkow** [comment]
There is one tool which may help you debug this issue (when you have repro described initially in this issue):
https://learn.microsoft.com/en-us/sysinternals/downloads/procmon

Then you should set filters: "Process Name" contains "vkl" + "Path" contains ".dll":

![image](https://github.com/openvkl/openvkl/assets/31285540/31feda6d-fdcb-4c49-8177-f8eee407bc66)

Then you'll be able to see from where it tries to look for those dll files.

**Comment by semeraro** [comment]
Thanks, I will check that out. I had been using dumpbin from the command line but this looks like a better option.

________________________________

**Comment by semeraro** [comment]
I did some experiments with dll shuffling. I found that if I swapped the rkcommon dll that I built from source for the rkcommon dll that comes with the openvkl binary distribution for windows the openvkl tests dont work. It seems I have done something wrong when building the rkcommon package.

**Comment by kraszkow** [comment]
Did you found solution? We recommend to build rkcommon using "superbuild", here you can see how it's done for older MSVC: 

https://github.com/openvkl/openvkl/blob/v1.3.2/.github/workflows/ci.windows.yml#L23

**Comment by semeraro** [comment]
Yes, I was finally able to compile openvkl and all of the dependencies. I first removed all traces of openvkl and dependencies from my laptop. I had become so confused about which version of which was where that I had to remove it all.
I then started with a fresh clone of openvkl 1.3.2. I proceeded with the superbuild. I did change a couple of things. I turned off parallel compile via the cmake settings in my vscode extensions. I have had problems with linux builds of things that trip over themselves when parallel builds are done. The other change I made was to the "default build command" in the superbuild cmakelists file. I changed the config from release to debug. I had been trying to make the change through the cmake settings but found some of the dependencies picked up on it and some did not. I wanted everything compiled debug in anticipation of having to do some tracing through everything. I left everything else alone. Configure and build went smoothly. And best of all the example applications worked fine. As has often been the case I cant say what it was I was doing wrong. I wish I could pinpoint it but I cant. Thanks for indulging me and for the help.

regards,
Dave
________________________________
